llvm/lib/Transforms/Utils/LoopUtils.cpp

```diff
@@ -1930,7 +1930,13 @@ llvm::hasPartialIVCondition(const Loop &L, unsigned MSSAThreshold,
   if (!TI || !TI->isConditional())
     return {};
 
-  auto *CondI = dyn_cast<CmpInst>(TI->getCondition());
+  Instruction *CondI = dyn_cast<CmpInst>(TI->getCondition());
+  if (!CondI) {
+    CondI = dyn_cast<TruncInst>(TI->getCondition());
+    if (CondI && CondI->getType() != Type::getInt1Ty(TI->getContext()))
```

Comment:
`CondI` must have type i1 here as it is used as condition of the branch, so no check should be necessary?

Maybe simpler to have


```
auto * CondI = dyn_cast<Instruction>(TI->getCondition());
if (!CondI || !isa<CmpInst, TruncInst>(CondI))
  return {}
```

Would be good to also document the restriction to compares & truncs. Perhaps it  would be best to remove the restriction if there is no strong reason to retain it?

Reply:
You've got a good point that it may be best to remove the restrictions, but I removed them and they broke quite a few tests, such as nested cases which shouldn't be unswitched and cases with phi nodes as the condition. Given this patch is a response to a specific regression I'd prefer to keep it simple if that's alright.

As for the CondI suggestion I've put that in and it looks cleaner - thanks. I've also added a short comment above to describe why it's just limited to Cmp and Trunc